plugins/device/rest.go

```diff
@@ -17,27 +17,38 @@ func deviceAddHandler(w http.ResponseWriter, r *http.Request) {
 
 	ctx := r.Context()
 	logger := gdctx.GetReqLogger(ctx)
+	peerID := mux.Vars(r)["peerid"]
+	if uuid.Parse(peerID) == nil {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, "Invalid peer-id passed in url")
+		return
+	}
 
 	req := new(deviceapi.AddDeviceReq)
 	if err := restutils.UnmarshalRequest(r, req); err != nil {
-		logger.WithError(err).Error("Failed to Unmarshal request")
-		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, "Unable to marshal request")
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, err)
 		return
 	}
-	peerID := mux.Vars(r)["peerid"]
-	if peerID == "" {
-		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, "peerid not present in request")
+
+	peerInfo, err := peer.GetPeer(peerID)
+	if err != nil {
+		logger.WithError(err).WithField("peerid", peerID).Error("Peer-id not found in store")
+		restutils.SendHTTPError(ctx, w, http.StatusNotFound, "Peer-id not found in store")
 		return
 	}
-	peerInfo, err := peer.GetPeer(peerID)
+
+	devices, err := CheckIfDeviceExist(req.Devices, peerInfo.Metadata["_devices"])
```

Comment:
All this must be done under a lock, including fetching from the store that you do above. See `glusterd2/commands/volumes/volume-start.go` for example.